Feature/upd build

tttomat19 opened the pull request:
- link to javadoc on gh-pages
- add copying of the final jar ./dcm4ceph-dist via antrun plugin

along with some cleanup changes mentioned in chat

Commit:
Remove v prefix from version number.

zgypa committed:
Update email, author and website in code header.

zgypa committed:
Renaming, ignore maven, install instructions.

zgypa committed:
Fix website.

zgypa committed:
fix char encoding.

zgypa committed:
Port make.sh to Makefile

zgypa committed:
Try to add CLI arguments, broken code.

zgypa committed:
Fix jar not building with dependencies.

zgypa committed:
Update version of commons-cli

zgypa committed:
Initial CLI implementation.

zgypa committed:
Remove redundant entrypoints.

zgypa committed:
Add missing import for Cephalogram

zgypa committed:
Add character encoding.

zgypa committed:
Refactor Java package to org.open_ortho

zgypa committed:
Update License to MIT.

zgypa committed:
Update instructions for building and running.

Commit:
fix java9+ issue with classcast

zgypa committed:
Fix missing age and study/series date.

zgypa committed:
Correct birthdate, add ethnic group, ref physician

zgypa committed:
Correct studydate and age.

zgypa committed:
Minor

Commit:
jar redist

Commit:
jar redist

Commit:
add link to javadoc hosted on github pages

Commit:
add link to javadoc hosted on github pages

zgypa commented on dcm4ceph-core/src/main/java/org/open_ortho/dcm4ceph/core/Cephalogram.java:
```diff
@@ -120,7 +129,7 @@ public void init() {
 
         // Load defaults first.
-        instanceProperties = FileUtils.loadProperties(this.getClass(),
+        instanceProperties = FileUtils.loadProperties(
```
Calling this class is no longer necessary i guess?

tttomat19 replied:
Agree, later properties are re-read in code. So it does not make sense to load at this point.
Ironically that was the code that was fixed in the first place :)

But let me please suggest another way, in commit [f5c2779bc00bc67d1e01ec9a69c7d2a83fa70be2](https://github.com/open-ortho/dcm4ceph/pull/6/commits/f5c2779bc00bc67d1e01ec9a69c7d2a83fa70be2)

The idea is foolproofness of program.
So even if user did not read the manual and did not provide .properties file one still will get output .dcm file.
The user will still get some results. And a warning.

zgypa replied:
@tttomat19 , you suggestion is good, but it would facilitate the generation of incorrect or incomplete DICOM files. If a DOB is unknown, you should probably set it null, or not add it in the DICOM header at all, rather than just adding a false default value of 0000-01-01. 

I think the default properties file can be used more as an example properties file than a default properties file. i guess i should create a ticket for this, to fix these up. We could, for example, pull out all the defaults from the code, and put them in the properties files for the different kind of input images. (Which is sort of leaning toward your initial suggestion of using existing tools, and wrapping them with appropriate XML data.).

I like the cleaning up of the redundant file readability check, but change the code to first look for the properties as filename with extension swapped with `.properties`, and if that fails throw an error with instructions for missing properties file and how to get an example of one (just like you did already). Also fix the the comments. I have added more comments about this individually.

tttomat19 replied:
Clear, reading ceph_defaults removed

zgypa commented on dcm4ceph-core/src/main/java/org/open_ortho/dcm4ceph/core/Cephalogram.java:
```diff
@@ -212,35 +221,45 @@ private void prepare() {
 
-        dcmobj.putString(Tag.SpecificCharacterSet, VR.CS, charset);
+        dcmobj.putString(Tag.SpecificCharacterSet, VR.CS, DEFAULT_CHARSET);
+        String transferSyntax = UID.JPEGBaseline1;
```
So you moved transferSyntax assignment here to be closer to its usage, but you kept DEFAULT_CHARSET at the top. Why not keep transferSyntax at top as well?

tttomat19 replied:
Not sure why I did that. Now this constant back at the top of the file.

zgypa commented on dcm4ceph-tool/pom.xml:
```diff
@@ -1,18 +1,89 @@
-  
+  <name>dcm4ceph-tool</name>
+  <version>1.0.0</version>
+  <url>http://dcm4cepch.antoniomagni.org</url>
```
This URL is incorrect and dead, should be current github repo https://github.com/open-ortho/dcm4ceph/

tttomat19 replied:
fixed

zgypa commented on mvnw:
```diff
@@ -0,0 +1,316 @@
+#!/bin/sh
```
Is adding boiler plate wrappers to each repo common Maven usage? Can you add the source from where you got these from?

tttomat19 replied:
These are simply created by the command `mvn wrapper:wrapper -Dmaven=3.8.6` to bootstrap maven wrapper initially.

I consider these files like a binary which is not for editing or reading.

After that one does not need to install maven at all, no M2_HOME, no adding mvn to path, no apt get install mvn, no maven version problems, because the version is defined by the source code in .mvn/maven-wrapper.properties file.

The idea is simpler redistribution. Overall this just lowers the entry bar to build the source a bit.

Still, one _can_ build the source with regular maven. I can add the instructions of how to do so. (Basically it is just `./mvnw` replaced with regular `mvn` binary no more than that).

zgypa replied:
i like these wrappers, i like how they worked instantly and i liked how they are generated from a maven target.

zgypa commented on mvnw.cmd:
```diff
@@ -0,0 +1,188 @@
+@REM ----------------------------------------------------------------------------
```
I assume this is boilerplate code as well. Can you add the source from where you got this from?

tttomat19 replied:
same as with ./mvnw, both files are created by one command

zgypa commented on pom.xml:
```diff
@@ -15,61 +15,17 @@
-    </repository>
-  </repositories>
-
   <distributionManagement>
```
This clearly needs to change, since we are no longer using this server anymore. What do you recommend setting these too? Is this even required at all?

tttomat19 replied:
Agree, removed. Unless the dcm4ceph-core is going to be published to some maven repo it is not required.

What we can do later is to publish jars to maven central or some other public maven repo, available for publishing by third parties. Maybe here? https://www.dcm4che.org/maven2/

zgypa reviewed (commented):
Overall i like what you did. Just had a few minor questions. See comments.

Commit:
foolproof working default properties

Commit:
foolproof working default properties

Commit:
transferSyntax static const moved up

Commit:
transferSyntax static const moved up

Commit:
fixed repo url

Commit:
fixed repo url

Commit:
rm obsolete distributionmanagement

Commit:
rm obsolete distributionmanagement

Commit:
Merge remote-tracking branch 'origin/feature/upd_build' into feature/upd_build

zgypa commented on dcm4ceph-core/src/main/java/org/open_ortho/dcm4ceph/core/Cephalogram.java:
```diff
@@ -107,12 +107,18 @@ public Cephalogram(File cephFile, File config) {
         init();
 
         if (config == null) {
```
Default should be looking for a properties file with same name as filename, with extension swapped with .properties. If that fails, then printing the message instructing user how to create properties file is appropriate. 

The reasoning is that you should not be able to generate a DICOM file without clinically relevant and mandatory DICOM tags. Remove and/or correct the comments in this section as appropriate

tttomat19 replied:
If I understand correctly, the option with explicit config file is to be removed.
Basically one and only way is the .properties file with swapped name. Just for overall simplicity.

zgypa replied:
No, what i meant was that the code first checks the CLI argument list: if properties is there, use that. If it is not, use the one with `.properties` file extension. If that fails, `Log.error` with instructions on how to create and use `.properties` files.

tttomat19 replied:
Returned the logic back, though CLI arg is for the properties not here. So I kinda decided this time not to introduce it unless told to do so :)

zgypa commented on dcm4ceph-core/src/main/java/org/open_ortho/dcm4ceph/core/Cephalogram.java:
```diff
@@ -285,6 +291,7 @@ public void setFromProperties(Properties cephprops) {
                                     + cephprops.getProperty("studyTime")));
         } catch (ParseException e) {
             e.printStackTrace();
+            Log.warn("using current datetime as studyDateTime");
```
Make log message consistent with previous one: "Could not parse Study Date Time correctly. Using current date time."

tttomat19 replied:
Done

zgypa commented on dcm4ceph-core/src/main/java/org/open_ortho/dcm4ceph/util/FileUtils.java:
```diff
@@ -169,20 +171,18 @@ public static Properties loadProperties(String conffile) {
+                    "Please use 2 files as input: %name%.jpg and %name%.properties \n" +
+                    "The DICOM output file will be created with default properties.\n" +
+                    "You may find example .properties file here: https://github.com/open-ortho/dcm4ceph/blob/master/dcm4ceph-sampledata/B1893F12.properties ");
+            return null;
```
Exit after Error message.

tttomat19 replied:
Done, not explicitly though. Not calling System.exit from utility file, but handling in function caller.

zgypa commented on dcm4ceph-core/src/main/java/org/open_ortho/dcm4ceph/util/FileUtils.java:
```diff
@@ -169,20 +171,18 @@ public static Properties loadProperties(String conffile) {
+            return props;
+        } catch (IOException e) {
+            e.printStackTrace();
+            Log.warn("Cannot read from file " + filename.toPath() + ".\n" +
```
Change to error and remove the line below about using the defaults properties. Default properties file should be the same file base name of the image name, but with .properties extension.

tttomat19 replied:
Done, added also reference to ceph_defaults as another example of .properties file.
One file is as real example, other as sensible defaults.

Commit:
javadoc build and release instructions

zgypa reviewed (commented):
Thanks for your suggestions. I will create a new issue for this.

zgypa commented on dcm4ceph-core/src/main/resources/ceph_defaults.properties:
```diff
@@ -8,7 +8,7 @@ patientID=""
 # Patient's Birth Date (0010,0030) 2 Birth date of the patient.
 # Use ISO format dd-MM-yyy 
-patientDOB=
+patientDOB=0000-01-01
```
Do not use a fake default for DOB. Better to keep Null, or not add DOB at all, if DOB is unk. Same with SID and SOD below.

tttomat19 replied:
Rolled back

Commit:
remove obsolete .sh scripts

Commit:
remove obsolete Makefile

Commit:
short circuit / fail fast style when cmd parse error, or when no file found

Commit:
remove default cfg loading, tmp remove explicit .properties file name, proper close resources , propagate exceptions up, only catch when able to handle.

Commit:
upd mag description

Commit:
upd .properties handling
add comments
upd error message
exit immediately when no .properties found
rolled back ceph_defaults

zgypa commented on dcm4ceph-core/src/main/java/org/open_ortho/dcm4ceph/core/Cephalogram.java:
```diff
@@ -323,8 +316,15 @@ public void setFromProperties(Properties cephprops) {
+        } catch (NumberFormatException e) {
+            e.printStackTrace();
+            Log.warn("Could not parse sid and sod. Please set proper sid= and sod= as decimal." +
+                    "\nDistances are not set.");
```
the second line is confusing. This warning would happen both when `sid` and `sod` are not set and when they are set to invalid values. So why are you printing "Distances are not set"? They might be set, but set to invalid values.

tttomat19 replied:
Removed, I mean they would not be set in the dicom tags.

zgypa commented on dcm4ceph-core/src/main/java/org/open_ortho/dcm4ceph/core/Cephalogram.java:
```diff
@@ -511,8 +511,8 @@ public String getDCMFileName() {
         float[] pixelspacing = getDXDetectorModule().getPixelSpacing();
-        for (int i = 0; i < pixelspacing.length; i++) {
-            if (pixelspacing[i] > getMaximumPixelSpacing()) {
+        for (float v : pixelspacing) {
```
oh cool! Was this already available in Java 8? I'm asking, since you suggested to maintain compatibility with Java8 compiler.

tttomat19 replied:
Yes, it is available in java 8

zgypa commented on dcm4ceph-tool/src/main/java/org/open_ortho/dcm4ceph/tool/ceph2dicomdir/Ceph2Dicom.java:
```diff
@@ -67,62 +67,73 @@ public static void main(String[] args) {
-			HelpFormatter formatter = new HelpFormatter();
-			formatter.printHelp("ceph2dicom", options);
+			// single file mode of operation
+			Log.info("Converting single TIF file to DCM.");
```
It's not converting TIF files yet. Just remove the TIF to read "converting single file to DCM".

tttomat19 replied:
Done

zgypa reviewed (commented):
I understand the urge of fixing things as you see them. But You again made a whole lotta changes to the code.... This delays the merging process by bringing up new issues all the time. At this stage, you will need to retest on Java8 and Java17 as you did initially, to make sure all the changes you made didn't break anything new.

I like the fact of cleaning up stuff. Clearly the code is poorly written, i didn't do a very good job back then. But i don't want you wasting your time doing things that might get re-done anyways. For example, we have a ticket to discuss porting to dcm4che4 or 5: if we decide to do it, this process might already force us to go through a lot of code and clean it up anyways.

Commit:
some fixes

Commit:
returned using explicit configFile in constructor

Commit:
upd comments

Commit:
fix compile

tttomat19 commented:
Basically re-tested java8 compile + java 17 run and vice-versa. All works fine now.
